File: greedy_algorithms/activity_selection.py
def select_activities(start, finish):
    results = [0]
    finished = finish[0]

    for i in range(1, len(start)):
        if finished <= start[i]:
            results.append(i)
            finished = finish[i]

    return results


if __name__ == '__main__':
    start = [1, 3, 0, 5, 3, 5, 6, 7, 8, 2, 12]
    finish = [4, 5, 6, 7, 9, 9, 10, 11, 12, 14, 16]

    print("  i |", end='')
    for i in range(len(start)):
        print(f"{i:>3}|", end='')
    print('')
    print("s[i]|", end='')
    for _s in start:
        print(f"{_s:>3}|", end='')
    print('')
    print("f[i]|", end='')
    for _f in finish:
        print(f"{_f:>3}|", end='')
    print('')

    activities = select_activities(start, finish)
    print('result = [', end='')
    for i, activity in enumerate(activities):
        print(activity, end=', ' if i < len(activities) - 1 else ']')
    print('')

File: greedy_algorithms/activity_selection.go
package main

import "fmt"

func selectActivity(s, f []int) []int {
	result := make([]int, 0)

	result = append(result, 0)
	finished := f[0]
	for idx := 1; idx < len(s); idx += 1 {
		if finished < s[idx] {
			result = append(result, idx)

			finished = f[idx]
		}
	}

	return result
}

func main() {
	s := []int{1, 3, 0, 5, 3, 5, 6, 8, 8, 2, 12}
	f := []int{4, 5, 6, 7, 9, 9, 10, 11, 12, 14, 16}

	fmt.Print("  i |")
	for i, _ := range s {
		fmt.Printf("%3d|", i)
	}
	fmt.Println()
	fmt.Print("s[i]|")
	for _, _s := range s {
		fmt.Printf("%3d|", _s)
	}
	fmt.Println()
	fmt.Print("f[i]|")
	for _, _f := range f {
		fmt.Printf("%3d|", _f)
	}
	fmt.Println()

	result := selectActivity(s, f)
	fmt.Println("activities =", result)
}
